accepted/install-locations.md

```diff
@@ -0,0 +1,239 @@
+* App-contained hosts (`apphost`, `comhost`, `ijwhost`, `nethost`) these are libraries which are shipped with the apps and thus fully owned/serviced by the app. They need to know how to find install locations to locate shared frameworks.
+
+The biggest problem is with the app-contained hosts which we can't service in any way. So introducing new install types or changing the default location can be breaking for apps using old hosts.
+
+The problem already exists since the `apphost` has been shipped in .NET Core 2.2 with support for framework dependent apps. The feature is used only occasionally. The most common case is probably dotnet global tools (`dotnet tool install -g <tool>`) which create framework dependent `apphost` for each installed tool.
+
+In .NET Core 3.0 framework dependent apps use an `apphost` by default, which makes this case far more prevalent. Scenarios requiring the other hosts are also likely to be relatively common (COM, IJW, native hosting). So the hosts which we ship with .NET Core 3.0 will be spread across all machines running .NET Core (not just development but production as well) without any ability to service them directly.
+
+This means that .NET Core 3.0 is the last chance to get the algorithms for install locations modified in "breaking ways". Future release will very likely have to be 100% backward compatible in this regard.
+
+
+## Proposal for 3.0
+
+### Globally registered install location :new:
+This is by far the most common case and in 3.0 we want to provide a way to globally install to a custom location. The typical case will still be to install into the default location, but custom location should be possible. Either location should behave the same and which one is used is solely decision of the installer.
+
+Note that the installer should try to avoid creating two global locations as only one can be registered. So if there is already some version installed it should only add to that location and create a new one. The only way to get two global locations should be to first install 3.0 into custom location and then install 2.* which always installs to the default location.
+
+To achieve the desired behavior the install location must be recorded in some system-wide configuration in a fixed location which is well-known.
+
+* Windows - the installer will store the install location in 32-bit registry (for both 32-bit and 64-bit installations):
+  * `HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\<arch>\InstallLocation` - where `<arch>` is one of the `x86`, `x64`, `arm` or `arm64`.
+  Note that this registry key is "redirected" that means that 32-bit processes see different copy of the key then 64bit processes. So it's important that both installers and the host access only the 32-bit view of the registry.
+* Linux and macOS - the installer will store the install location in a text file
+  * `/etc/dotnet/install_location.conf` - the file will contain a single line which is the path.
```

Comment:
Would it be better if we put more settings than just a single one per file? There are other settings that should go into `/etc/dotnet` (think any env var we have today). Having one file per setting seems excessive.

Reply:
It would be interesting to know which variables might be useful. The only ones which might make at least some sense to define globally would probably be:
* `DOTNET_RUNTIME_ID`
* `DOTNET_SHARED_STORE` - we're trying to move away from this.
* `DOTNET_MULTILEVEL_LOOKUP` - this has no effect on Linux

We could change it for the sake future extensibility:
File name: `install.conf`   (better ideas? Trying to reflect what we have on Windows in registry at least a little bit for consistency. If needed we could add installed SDKs into it, and the `hostfxr` version)

Content in the `.ini` format:
```ini
#comment
InstallLocation=<path>
```

For now we would probably implement a VERY simplistic .ini parser, just ignore comments and split on `=` sign.

Reply:
Basically anything that is an environment variable check today:

* DOTNET_CLI_TELEMETRY_OPTOUT
* DOTNET_SKIP_FIRST_TIME_EXPERIENCE
* DOTNET_DISABLE_MULTICOREJIT
* DOTNET_CLI_UI_LANGUAGE

See also - https://github.com/dotnet/cli/issues/2960

Reply:
Good point - I didn't realize that this would apply to SDK as well.
What about the filename/format, does that sound reasonable?

Reply:
I think that file name and format sounds reasonable. Just be sure to allow spaces around the `=` as well:

```ini
InstallLocation = <path>
```

Reply:
cc @nguerrera what to think on the implications for an eventual dotnet config verb? Seems to me like there are some overlap here.

Reply:
If people will have a negative reaction to other config switches going in another ini file next to this one, I think we should consider going back to just the install location here. I don't want to end up in a place where we can't evolve the full set of settings because it will break apphosts. It makes sense to me to keep the part that the apphost will read down to the bare minimum. I think a file with a path is better for that.

Reply:
In other words, I agree 1 setting per file is excessive, but keeping the settings that are read by unserviceable apphost code separate seems wise.

Reply:
@eerhardt would you be OK with the approach proposed by @nguerrera?
Keep the things which are required to run apps separate from everything else and as simple as possible...
(in which case I would also like to keep the simple text file with first line meaning the location, everything else ignored for now).

Reply:
If we go this way, I would consider using a less suggestive name than the .conf extension: `/etc/dotnet/install_location`. The .conf extension and the lack of location are suggestive that we would put more things in there, but I think that is probably the only thing the apphost will read. The rest can be read from code we can service in a different file, right?

Reply:
Another possible setting that could go in this file is the `DOTNET_RUNTIME_ID` setting, in cases where a user needs to override this and can't/won't use an environment variable.

Looking for more cases of the host reading environment variables:

https://github.com/dotnet/core-setup/search?q=getenv&unscoped_q=getenv

`CORE_BREADCRUMBS` and `CORE_SERVICING` could be others.


I'm OK with having a split between the "runtime" config and "SDK" config, especially considering if the SDK config needs a more complex format. But from a customer's perspective, it feels a bit arbitrary, and I don't know if all users understand the difference.

I think my biggest concern is not having 1 file per setting. But I agree we don't need all settings in one file.

Reply:
Even for the runtime/host settings, there's a distinction between what must be read by apphost code sitting in customer app and what can be read by serviceable code. This is why I think we should keep the thing read by apphost distinct and as simple as possible.